packages/mongodb-explain-compat/lib/index.js

```diff
@@ -72,18 +72,143 @@ function mapStages(queryPlan, sbeExecutionStages) {
+ *
+ * @param {Object} explain Raw explain object.
+ * @returns Classic explain object
+ */
+module.exports = function (explain) {
   explain = JSON.parse(JSON.stringify(explain));
-  if ((explain.explainVersion || 0) < 2) {
+
+  // In a sharded aggregation, we don't have explainVersion :(
+  // In a sharded find, its represented by explain.queryPlanner.mongosPlannerVersion
+
+  // return explain that uses classic engine (except for sharded response)
+  if (explain.explainVersion && explain.explainVersion < 2) {
     return explain;
   }
   delete explain.explainVersion;
-  explain.queryPlanner.plannerVersion = 1;
-  const winningPlan = explain.queryPlanner.winningPlan.queryPlan;
-  explain.queryPlanner.winningPlan = winningPlan;
 
-  explain.executionStats.executionStages =
-    mapStages(winningPlan, explain.executionStats.executionStages);
+  if (isAggregationExplain(explain)) {
+    if (isShardedAggregationExplain(explain)) {
+      return mapShardedAggregation(explain);
+    }
+    return mapsUnshardedAggregation(explain);
+  }
 
-  return JSON.parse(JSON.stringify(explain));
+  // Sharded Find
+  if (isShardedFind(explain)) {
+    return mapShardedFind(explain);
+  }
+
+  // Unsharded Find
+  return mapUnshardedFind(explain);
 };
+
+function mapsUnshardedAggregation(explain) {
+  if (!explain.stages || explain.stages.length === 0) {
+    return explain;
+  }
+  const stages = explain.stages.map((stage) => {
+    if (!isCursorStage(stage)) {
+      return stage;
+    }
+    const stageKey = getStageCursorKey(stage);
+    stage[stageKey] = _mapPlannerStage(stage[stageKey]);
+    return stage;
+  });
+  explain.stages = stages;
+  return explain;
+}
+function mapShardedAggregation(explain) {
+  if (!explain.shards) {
+    return explain;
+  }
+  const shards = {};
+  for (const shardName in explain.shards) {
+    // Shard with stages
+    if (explain.shards[shardName].stages) {
+      shards[shardName] = mapsUnshardedAggregation(explain.shards[shardName]);
+    } else {
+      shards[shardName] = mapUnshardedFind(explain.shards[shardName]);
+    }
+  }
+  explain.shards = shards;
+  return explain;
+}
+
+function mapShardedFind(explain) {
+  const queryPlanner = explain.queryPlanner;
+  const executionStats = explain.executionStats;
+  if (!queryPlanner.winningPlan.shards) {
+    return explain;
+  }
+  queryPlanner.winningPlan.shards.forEach((shard, index) => {
+    const winningPlan = shard.winningPlan.queryPlan;
+    if (winningPlan) {
+      const executionStages = mapStages(
+        winningPlan,
+        executionStats.executionStages.shards[index].executionStages
+      );
+
+      queryPlanner.winningPlan.shards[index].winningPlan = winningPlan;
+      executionStats.executionStages.shards[index].executionStages =
+        executionStages;
+    }
+  });
+  explain.queryPlanner = queryPlanner;
+  explain.executionStats = executionStats;
+  return JSON.parse(JSON.stringify(explain));
```

Comment:
We already have that in place at main export. This, I think has been there to clear out all the information which is not needed like the undefined props, symbols which are added while mapping.

Reply:
Then maybe I’d suggest keeping it there for consistency/to avoid duplicating this bit every time, since it’s part of the external interface of this package and not part of each function’s individual functionality?